subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformationChain.java

```diff
@@ -18,27 +18,51 @@
 
 import org.gradle.api.Action;
 
+import javax.annotation.Nullable;
+
 /**
  * A series of {@link TransformationStep}s.
  */
 public class TransformationChain implements Transformation {
 
-    private final Transformation first;
-    private final Transformation second;
+    private final Transformation init;
+    private final TransformationStep last;
     private final int stepsCount;
 
-    public TransformationChain(Transformation first, Transformation second) {
-        this.first = first;
-        this.second = second;
-        this.stepsCount = first.stepsCount() + second.stepsCount();
+    private static final Transformation EMPTY = new Transformation() {
```

Comment:
❓ Do we really need this instance?

It seems we are guarding against it in both `getInit()` and in `toString()`, and we are using it in the `visitTransformationSteps` and `requiresDependencies`. 

We can remove the instance and add the null checks for the field where needed. IMO, it's easier to guard against nullability that against a specific singleton instance.

Additionally, this anonymous implementation seems to be the only reason to have the `Transformation` interface. Otherwise, `TransformationChain` will be the only implementation, and the interface can be removed all together, which I am in favor of.